### PR TITLE
WiP(invite-ui): Initial move of invite UI to invite button

### PR DESCRIPTION
**Pull request by yanas**
This is still WiP and still needs some CSS love before we'll be able to merge it.

### CODE REVIEW

File: react/features/video-quality/components/VideoQualityLabel.web.js [hunk 1]
```diff
@@ -219,7 +207,6 @@ function _mapStateToProps(state) {
         _audioOnly: audioOnly,
         _conferenceStarted: Boolean(conference),
         _filmstripVisible: visible,
-        _remoteVideosVisible: shouldRemoteVideosBeVisible(state),
```
Comment (virtuacoplenny) [inline react/features/video-quality/components/VideoQualityLabel.web.js]: Could you do me a big favor and also remove this logic from recording label? I had duplicated the logic for recording label reactification, with the intent to refactor later, but moving the local thumbnail up basically saves me a lot of work.

File: css/_vertical_filmstrip_overrides.scss [hunk 2]
```diff
@@ -161,7 +172,7 @@
     .video-state-indicator.moveToCorner {
         transition: right 0.5s;
 
-        &.with-filmstrip.with-remote-videos {
```
Comment (virtuacoplenny) [inline css/_vertical_filmstrip_overrides.scss]: The greatest danger I see with comments: getting outdated. The comments about the with-remote-videos and without-remove-videos classes can be deleted too.

File: react/features/invite/components/InviteButton.web.js [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+        if (props._isDialOutAvailable) {
+            inviteItems.splice(0, 0, {
+                content: t('dialOut.dialOut'),
+                action: () => this.props.openDialog(DialOutDialog)
+            });
+        }
+
+        if (props._isAddToCallAvailable) {
+            inviteItems.splice(0, 0, {
```
Comment (virtuacoplenny) [inline react/features/invite/components/InviteButton.web.js]: Why did you choose to do splice instead of unshift or maybe initializing inviteItems as empty first?

File: lang/main.json [hunk 4]
```diff
@@ -440,8 +440,10 @@
         "hidePassword": "Hide password",
         "inviteTo": "Invite people to __conferenceName__",
         "invitedYouTo": "__userName__ has invited you to the __inviteURL__ conference",
+        "invitePeople": "Invite people",
         "locked": "This call is locked. New callers must have the link and enter the password to join.",
         "showPassword": "Show password",
+        "shareTheLink": "Share the link",
```
Comment (virtuacoplenny) [inline lang/main.json]: I noticed this same copy being used for 4 different translation keys (one of them no longer referenced). Is there any guideline for this situation to prevent duplication? Is it an occurrence worth addressing? Should all the other usages be moved over to this one?

File: css/_vertical_filmstrip_overrides.scss [hunk 5]
```diff
@@ -50,9 +50,20 @@
...
+        /**
+         * Re-styles the local Video and invite button to better fit the
+         * vertical filmstrip layout.
+         */
         #filmstripLocalVideo {
+            bottom: 5px;
+            flex-direction: column-reverse;
             height: auto;
-            justify-content: flex-end;
+            justify-content: flex-start;
```
Comment (virtuacoplenny) [inline css/_vertical_filmstrip_overrides.scss]: You'll probably have to adjust the local stats and maybe remote stats dialogs. On the plus, it might end up looking better now.

File: react/features/invite/components/InviteButton.web.js [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+        /**
+         * The function closing the dialog.
+         */
+        openDialog: React.PropTypes.func,
+
+        /**
+         * Invoked to obtain translated strings.
+         */
+        t: React.PropTypes.func
```
Comment (virtuacoplenny) [inline react/features/invite/components/InviteButton.web.js]: I saw in another PR the pattern of storing PropTypes in an intermediary variable. I think that should be standard now because React.PropTypes is being deprecated.

File: react/features/invite/components/InviteButton.web.js [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+         */
+        _isAddToCallAvailable: React.PropTypes.bool,
+
+        /**
+         * Indicates if the "Dial out" feature is available.
+         */
+        _isDialOutAvailable: React.PropTypes.bool,
+
+        /**
+         * The function closing the dialog.
```
Comment (virtuacoplenny) [inline react/features/invite/components/InviteButton.web.js]: I think the function opens the dialogs.

File: react/features/invite/components/InviteButton.web.js [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+    /**
+     * Handles selection of the invite options.
+     *
+     * @param { Object } option - The invite option that has been selected from
+     * the dropdown menu.
+     * @private
+     * @returns {void}
+     */
+    _onInviteOptionSelected(option) {
+        this.state.inviteOptions[0].items.map(item => {
```
Comment (virtuacoplenny) [inline react/features/invite/components/InviteButton.web.js]: Why are you using map if the returned value isn't returning anything? Is it necessary to even have iteration logic? I'm assuming you can even have the action within the option.

File: react/features/invite/components/InviteButton.web.js [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+ * }}
+ */
+function _mapStateToProps(state) {
+    const { enableUserRolesBasedOnToken } = state['features/base/config'];
+
+    const { conference } = state['features/base/conference'];
+
+    const isDialOutButtonEnabled
+        = interfaceConfig.TOOLBAR_BUTTONS.indexOf(DIAL_OUT_NAME) !== -1
+        || interfaceConfig.MAIN_TOOLBAR_BUTTONS.indexOf(DIAL_OUT_NAME) !== -1;
```
Comment (virtuacoplenny) [inline react/features/invite/components/InviteButton.web.js]: There is an isButtonEnabled helper for this check.

File: lang/main.json (hunk 4 above)
Comment (yanas) [reply lang/main.json]: Good catch. The translation file isn't currently structured in the best way to make it easy to avoid duplications. I guess the basic rule is to search for your string before adding one. I generally do that but have missed this one. I'll fix this.

File: react/features/invite/components/InviteButton.web.js (hunk 8 above)
Comment (yanas) [reply react/features/invite/components/InviteButton.web.js]: I was using it differently before and now it's not needed anymore.

File: react/features/invite/components/InviteButton.web.js (hunk 6 above)
Comment (yanas) [reply react/features/invite/components/InviteButton.web.js]: I saw we already import prop-types in AbstractApp.js, so I'll start to use this one directly.

File: react/features/invite/components/InviteButton.web.js (hunk 3 above)
Comment (yanas) [reply react/features/invite/components/InviteButton.web.js]: I don't remember anymore :)

File: react/features/toolbox/defaultToolbarButtons.js [hunk 10]
```diff
@@ -18,19 +17,6 @@ declare var JitsiMeetJS: Object;
  * All toolbar buttons' descriptors.
  */
 const buttons: Object = {
-    addtocall: {
-        classNames: [ 'button', 'icon-add' ],
```
[the remark this answers is not in the document]
Comment (yanas) [reply react/features/toolbox/defaultToolbarButtons.js]: Yep, still need to fix that!

**Comment by yanas** [comment]
PR is not ready for merge. I continue to work on 2 tasks. First one is to remove unused icons. The second one is to move addtocall, dialout and invite out of the toolbar array in interfaceConfig to a separate INVITE_OPTIONS list.

File: css/_filmstrip.scss [hunk 11]
```diff
@@ -48,7 +48,6 @@
     &__videos {
         @extend %align-right;
         position:relative;
-        height:196px;
```
Comment (yanas) [inline css/_filmstrip.scss]: Not sure why this was here, but it was messing with my new invite button UI, so I removed it. Tested in horizontal and vertical filmstrip mode and it didn't seem to have any consequence.

File: react/features/invite/components/InviteButton.web.js (hunk 3 above)
Comment (yanas) [reply react/features/invite/components/InviteButton.web.js]: Now I changed the code I little bit and I use splice in order to position the items in the list in the way they have been defined in interfaceConfig.

File: css/_filmstrip.scss (hunk 11 above)
Comment (virtuacoplenny) [reply css/_filmstrip.scss]: Sounds good. We shouldn't be afraid to kill off CSS because we can likely fix it easy enough.

File: react/features/invite/components/InviteButton.web.js [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+        const { VERTICAL_FILMSTRIP } = interfaceConfig;
+
+        return (
+            <div className = 'filmstrip__invite'>
+                <div className = 'invite-button-group'>
+                    <Button
+                        onClick = { this._onInviteClick }
+                        shouldFitContainer = { true }>
+                        { t('invite.invitePeople') }
+                    </Button>
```
Comment (virtuacoplenny) [inline react/features/invite/components/InviteButton.web.js]: You might want to add truncation of the text if possible with Atlaskit, in case some translation makes the button stretch horizontally. Maybe use max width on the containing element as a backup option for preventing horizontal stretch?

File: css/_vertical_filmstrip_overrides.scss [hunk 13]
```diff
@@ -69,10 +80,11 @@
...
             flex-direction: column;
             height: auto;
-            overflow-x: hidden !important;
+            justify-content: flex-end;
 
             .remote-videos-container {
-                flex-direction: column;
+                flex-direction: column-reverse;
+                overflow-x: hidden !important;
             }
```
Comment (virtuacoplenny) [inline css/_vertical_filmstrip_overrides.scss]: What let you have vertical filmstrip scrolling start from the bottom of the filmstrip?

File: css/_vertical_filmstrip_overrides.scss (hunk 5 above)
Comment (virtuacoplenny) [reply css/_vertical_filmstrip_overrides.scss]: I think the remote video menu needs adjusting now too... Because remote videos start on the bottom in vertical filmstrip, that menu can get cut off if there are many options, which I can see happening with mute, kick, local volume, and remote control.

File: css/_vertical_filmstrip_overrides.scss [hunk 14]
```diff
@@ -69,10 +80,11 @@
             flex: 1;
             flex-direction: column;
             height: auto;
-            overflow-x: hidden !important;
+            justify-content: flex-end;
 
             .remote-videos-container {
-                flex-direction: column;
+                flex-direction: column-reverse;
+                overflow-x: hidden !important;
```
Comment (virtuacoplenny) [inline css/_vertical_filmstrip_overrides.scss]: I don't think the !important is needed. In general !importants aren't needed for self-controlled css and they're tricky to workaround, although it isn't likely to become a problem. I still feel it's better to avoid the finality of !important. Instead BEM or increasing specificity might be warranted.

File: interface_config.js [hunk 15]
```diff
@@ -35,13 +35,16 @@ var interfaceConfig = { // eslint-disable-line no-unused-vars
...
         //extended toolbar
-        'profile', 'addtocall', 'contacts', 'info', 'chat', 'recording', 'etherpad', 'sharedvideo', 'dialout', 'settings', 'raisehand', 'videoquality', 'filmstrip'], // jshint ignore:line
+        'profile', 'contacts', 'info', 'chat', 'recording', 'etherpad', 'sharedvideo', 'settings', 'raisehand', 'videoquality', 'filmstrip'], // jshint ignore:line
     /**
      * Main Toolbar Buttons
      * All of them should be in TOOLBAR_BUTTONS
      */
     MAIN_TOOLBAR_BUTTONS: ['microphone', 'camera', 'desktop', 'invite', 'fullscreen', 'fodeviceselection', 'hangup'], // jshint ignore:line
     SETTINGS_SECTIONS: ['language', 'devices', 'moderator'],
+    // TODO: rename invite to sharelink, once the complete transition to the
```
Comment (virtuacoplenny) [inline interface_config.js]: Is this still a todo? Does it seem out of scope to do it now?

File: react/features/filmstrip/components/Filmstrip.web.js [hunk 16]
```diff
@@ -107,7 +110,12 @@ class Filmstrip extends Component {
                         className = 'filmstrip__videos'
                         id = 'filmstripLocalVideo'
                         onMouseOut = { this._onMouseOut }
-                        onMouseOver = { this._onMouseOver } />
+                        onMouseOver = { this._onMouseOver }>
+                        {
+                            interfaceConfig.filmStripOnly
+                                ? null : <InviteButton />
+                        }
+                    </div>
```
Comment (virtuacoplenny) [inline react/features/filmstrip/components/Filmstrip.web.js]: I'm feeling like this could be a prop instead, be it a new one or an existing one. I think this.props.displayToolbox is already passed in for filmstrip only too. Reducing the usage of the global interfaceConfig might be warranted in general when configurations can be passed in, that way component assume less about the world around them and become more encapsulated.

File: react/features/invite/components/InviteButton.web.js [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+            );
+        }
+
+        this.state = {
+            /**
+             * The list of invite options.
+             */
+            inviteOptions: [
+                {
+                    items: inviteItems
```
Comment (virtuacoplenny) [inline react/features/invite/components/InviteButton.web.js]: Is this state necessary? Could this be done within the scope of render?

File: react/features/invite/components/InviteButton.web.js [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+        _isDialOutAvailable:
+            getLocalParticipant(state).role === PARTICIPANT_ROLE.MODERATOR
+            && conference && conference.isSIPCallingSupported()
+            && isInviteOptionEnabled(DIAL_OUT_OPTION)
+            && (!enableUserRolesBasedOnToken || !isGuest)
+    };
+}
+
+export default translate(connect(
+    _mapStateToProps, { openDialog })(InviteButton));
```
Comment (virtuacoplenny) [inline react/features/invite/components/InviteButton.web.js]: I've seen you use this pattern before where you map an action to a prop. What are the advantages of this approach?

File: react/features/invite/functions.js [hunk 19]
```diff
@@ -77,3 +78,27 @@ export function inviteRooms(conference, rooms) {
...
+/**
+ * Get the position of the invite option in the interfaceConfig.INVITE_OPTIONS
+ * list.
+ *
+ * @param {string} optionName - The invite option name.
+ * @private
+ * @returns {number} - The position of the option in the list.
+ */
+export function getInviteOptionPosition(optionName) {
+    return interfaceConfig.INVITE_OPTIONS.indexOf(optionName);
```
Comment (virtuacoplenny) [inline react/features/invite/functions.js]: What's your thought on this: if helper functions are used in one component only, should they put into functions.js?

File: css/_filmstrip.scss [hunk 20]
```diff
@@ -48,21 +48,64 @@
...
+        &#filmstripLocalVideo {
+            bottom: 32px;
+            flex-direction: column;
+
+            /**
+             * The invite button style.
+             */
+            .filmstrip__invite {
+                padding-bottom: 5px;
+                z-index: $dropdownZ;
```
Comment (virtuacoplenny) [inline css/_filmstrip.scss]: In horizontal filmstrip, local video tooltips and stats menu display below the invite button. I'm not sure what z-index wars will need to be fought to fix it...

File: react/features/invite/components/InviteButton.web.js [hunk 21]
```diff
@@ -0,0 +1,231 @@
...
+        if (props._isAddToCallAvailable) {
+            inviteItems.splice(
+                getInviteOptionPosition(ADD_TO_CALL_OPTION),
+                0,
+                {
+                    content: interfaceConfig.ADD_PEOPLE_APP_NAME,
+                    action: () => this.props.openDialog(AddPeopleDialog)
+                }
+            );
+        }
```
Comment (virtuacoplenny) [inline react/features/invite/components/InviteButton.web.js]: Why prefer splice over doing a push? Check for featureAvailable, push to array.

File: react/features/invite/components/InviteButton.web.js [hunk 22]
```diff
@@ -0,0 +1,231 @@
...
+                    { this.props._isDialOutAvailable
+                        || this.props._isAddToCallAvailable
+                        ? <DropdownMenu
+                            items = { this.state.inviteOptions }
+                            onItemActivated = { this._onInviteOptionSelected }
+                            position = { VERTICAL_FILMSTRIP
+                                ? 'bottom right'
+                                : 'top right' }
+                            shouldFlip = { true }
+                            triggerType = 'button' />
```
Comment (virtuacoplenny) [inline react/features/invite/components/InviteButton.web.js]: What do you think about the arrow in the dropdown in horizontal filmstrip mode? It currently points down but the dropdown opens above.

File: react/features/invite/components/InviteButton.web.js (hunk 18 above)
Comment (yanas) [reply react/features/invite/components/InviteButton.web.js]: I like it because I don't need to call dispatch and it shortens my code when calling the action.

File: interface_config.js (hunk 15 above)
Comment (yanas) [reply interface_config.js]: We're going to get rid of the whole button, so I'll remove the todo.

File: react/features/filmstrip/components/Filmstrip.web.js (hunk 16 above)
Comment (yanas) [reply react/features/filmstrip/components/Filmstrip.web.js]: I'll check the displayToolbox prop. I also considered a prop, but my thinking was that once we put interfaceConfig in the store it will be easy to replace all calls. However, it will also mean that most of our components are redux aware even when probably not needed, so yeah I'll see if I can replace it with a prop.

File: react/features/invite/components/InviteButton.web.js (hunk 22 above)
Comment (yanas) [reply react/features/invite/components/InviteButton.web.js]: Do you know any property of the AtlasKit dropdown that can fix this? I personally thought that I can't change this.

File: modules/UI/videolayout/Filmstrip.js [hunk 23]
```diff
@@ -433,11 +423,14 @@ const Filmstrip = {
             promises.push(new Promise((resolve) => {
                 // Let CSS take care of height in vertical filmstrip mode.
                 if (interfaceConfig.VERTICAL_FILMSTRIP) {
-                    resolve();
+                    $('#filmstripLocalVideo').animate({
+                        // adds 4 px because of small video 2px border
+                        width: local.thumbWidth + 4
```
Comment (yanas) [inline modules/UI/videolayout/Filmstrip.js]: Adding 4px here is quite shady, but we were already doing it below for the horizontal filmstrip, so I decided it's not that bad. The below code already got out of sync with the css though, so maybe getting the border from the component will be better.

File: css/_filmstrip.scss (hunk 20 above)
Comment (virtuacoplenny) [reply css/_filmstrip.scss]: Okay, so the battle is with .hide-videos setting a z-index that's too low and removing the z-index might be a naive yet effective fix. I can't remember immediately why the lower z-index is needed, maybe to prevent the label from being unclickable when docked to the top right of the screen. Regardless I doubt the z-index is needed now because the interaction between video label and filmstrip is much simpler with the local video on top.

File: css/_vertical_filmstrip_overrides.scss (hunk 13 above)
Comment (virtuacoplenny) [reply css/_vertical_filmstrip_overrides.scss]: I'm noticing an issue right now on firefox with vertical filmstrip where the videos do not scroll and instead keep stacking on each other until they're pushing out of the window. I'm wondering if it's related to the style changes to get the scrolling to start from the bottom of the filmstrip.

**Comment by virtuacoplenny** [comment]
My commit for feat(connection-indicator) is included in the changes. Not sure why. It'll be hidden when squashing but I want to confirm it existing in your change set will not cause issues.

**Comment by yanas** [comment]
Yeah, I also noticed and I'm not sure why. I rebase to master every time and force push.. It doesn't pose any problems for me, so I just left it this way.

File: react/features/invite/components/InviteButton.web.js (hunk 21 above)
Comment (yanas) [reply react/features/invite/components/InviteButton.web.js]: Because I want to insert the element at the specific position defined in interfaceConfig.

File: react/features/invite/components/InviteButton.web.js (hunk 22 above)
Comment (virtuacoplenny) [reply react/features/invite/components/InviteButton.web.js]: In atlaskit using their default dropdown menu? No, I don't see a property. You can use a custom icon/trigger but it's probably not worthwhile. It's cold to say but it's an atlaskit issue because even in their storybook the arrow doesn't always point to where the dropdown opens.

File: react/features/invite/components/InviteButton.web.js (hunk 21 above)
Comment (virtuacoplenny) [reply react/features/invite/components/InviteButton.web.js]: Ah okay, I'm seeing it now. I thought you were just always splicing to the front. It makes sense then. An alternative might be to map or filter over the array and maybe keep a map somewhere of expected invite option and its respective enable check as well as the dropdown option.

File: css/_vertical_filmstrip_overrides.scss (hunk 13 above)
Comment (yanas) [reply css/_vertical_filmstrip_overrides.scss]: I'll check that. The thing I did to make the scroll work from bottom is to set column-reverse to the inner container and move the overflow-x from the parent to the inner container as well. The set the flex-end to the outer container.

File: react/features/invite/components/InviteButton.web.js (hunk 17 above)
Comment (yanas) [reply react/features/invite/components/InviteButton.web.js]: By using the state I can make sure that "inviteItems" array is only updated when one of the 2 properties I check in componentWillReceiveProps changes. This way the dropdown will be re-rendered only if the inviteItems has changed.

Comment (virtuacoplenny) [reply react/features/invite/components/InviteButton.web.js]: Do you feel that optimization is necessary? What about using PureComponent then?

File: react/features/invite/functions.js (hunk 19 above)
Comment (yanas) [reply react/features/invite/functions.js]: I would agree in general. However, I'm not sure because I want to keep this one together with the isInviteOptionEnabled, which is used in the mapStateToProps, so if I move them in InviteButton, I'll need to put at least this one out of the component so that I can use it in the mapStateToProps. It looks to me cleaner to have them in functions, but it's not a strong opinion.

File: react/features/invite/components/InviteButton.web.js (hunk 22 above)
Comment (yanas) [reply react/features/invite/components/InviteButton.web.js]: Yeah, I would leave that one to them and report it there.

File: react/features/invite/components/InviteButton.web.js (hunk 17 above)
Comment (yanas) [reply react/features/invite/components/InviteButton.web.js]: Actually, I'm not sure I see what bothers you in here. I feel like this is the most straightforward react way to do such a thing.

File: react/features/invite/components/InviteButton.web.js (hunk 21 above)
Comment (yanas) [reply react/features/invite/components/InviteButton.web.js]: Is splice that bad? :)

Comment (virtuacoplenny) [reply react/features/invite/components/InviteButton.web.js]: No it's not bad.

File: react/features/invite/components/InviteButton.web.js (hunk 17 above)
Comment (virtuacoplenny) [reply react/features/invite/components/InviteButton.web.js]: I generally prefer reducing the usage of member variables if it's possible to avoid them without much effort. It comes down to keeping variable access as restrictive as possible.
